File: app/src/main/java/org/wikipedia/page/references/ReferenceDialog.java
package org.wikipedia.page.references;

import android.app.Dialog;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewpager2.widget.ViewPager2;

import com.google.android.material.bottomsheet.BottomSheetBehavior;
import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.google.android.material.tabs.TabLayout;
import com.google.android.material.tabs.TabLayoutMediator;

import org.wikipedia.R;
import org.wikipedia.activity.FragmentUtil;
import org.wikipedia.page.ExtendedBottomSheetDialogFragment;
import org.wikipedia.page.LinkHandler;
import org.wikipedia.page.LinkMovementMethodExt;
import org.wikipedia.util.DimenUtil;
import org.wikipedia.util.StringUtil;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

import static org.wikipedia.util.L10nUtil.setConditionalLayoutDirection;

/**
 * A dialog that displays the currently clicked reference.
 */
public class ReferenceDialog extends ExtendedBottomSheetDialogFragment {
    public interface Callback {
        LinkHandler getLinkHandler();
        List<PageReferences.Reference> getReferencesGroup();
        int getSelectedReferenceIndex();
    }

    @BindView(R.id.reference_pager) ViewPager2 referencesViewPager;
    @BindView(R.id.page_indicator_view) TabLayout pageIndicatorView;
    @BindView(R.id.indicator_divider) View pageIndicatorDivider;
    @BindView(R.id.reference_title_text) TextView titleTextView;
    private Unbinder unbinder;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_references_pager, null);
        unbinder = ButterKnife.bind(this, rootView);
        if (callback() == null || callback().getReferencesGroup() == null) {
            dismiss();
            return rootView;
        }

        titleTextView.setText(requireContext().getString(R.string.reference_title, ""));
        referencesViewPager.setOffscreenPageLimit(2);
        referencesViewPager.setAdapter(new ReferencesAdapter(callback().getReferencesGroup()));
        new TabLayoutMediator(pageIndicatorView, referencesViewPager, (tab, position) -> { }).attach();
        referencesViewPager.setCurrentItem(callback().getSelectedReferenceIndex(), true);

        setConditionalLayoutDirection(rootView, callback().getLinkHandler().getWikiSite().languageCode());
        return rootView;
    }

    @Override
    public void onStart() {
        super.onStart();
        if (callback() != null && callback().getReferencesGroup() != null && callback().getReferencesGroup().size() == 1) {
            pageIndicatorView.setVisibility(View.GONE);
            pageIndicatorDivider.setVisibility(View.GONE);
        } else {
            BottomSheetBehavior<View> behavior = BottomSheetBehavior.from((View) getView().getParent());
            behavior.setPeekHeight(DimenUtil.getDisplayHeightPx() / 2);
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    @NonNull
    private String processLinkTextWithAlphaReferences(@NonNull String linkText) {
        boolean isLowercase = linkText.contains("lower");
        if (linkText.contains("alpha ")) {
            String[] strings = linkText.split(" ");
            String alphaReference = StringUtil.getBase26String(Integer.parseInt(strings[strings.length - 1].replace("]", "")));
            alphaReference = isLowercase ? alphaReference.toLowerCase() : alphaReference;
            linkText = alphaReference;
        }
        return linkText.replace("[", "").replace("]", "") + ".";
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        return new BottomSheetDialog(requireActivity(), getTheme()){
            @Override
            public void onBackPressed() {
                if (referencesViewPager.getCurrentItem() > 0) {
                    referencesViewPager.setCurrentItem(referencesViewPager.getCurrentItem() - 1, true);
                } else {
                    super.onBackPressed();
                }
            }
        };
    }

    private class ViewHolder extends RecyclerView.ViewHolder {
        private TextView pagerReferenceText;
        private TextView pagerIdText;

        ViewHolder(View itemView) {
            super(itemView);
            pagerReferenceText = itemView.findViewById(R.id.reference_text);
            pagerReferenceText.setMovementMethod(new LinkMovementMethodExt(callback().getLinkHandler()));
            pagerIdText = itemView.findViewById(R.id.reference_id);
        }

        void bindItem(CharSequence idText, CharSequence contents) {
            pagerIdText.setText(idText);
            pagerReferenceText.setText(contents);
        }
    }

    private class ReferencesAdapter extends RecyclerView.Adapter {
        private List<PageReferences.Reference> references = new ArrayList<>();

        ReferencesAdapter(@NonNull List<PageReferences.Reference> adjacentReferences) {
            references.addAll(adjacentReferences);
        }

        @Override
        public int getItemCount() {
            return references.size();
        }

        @NonNull
        @Override
        public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            LayoutInflater inflater = LayoutInflater.from(getContext());
            View view = inflater.inflate(R.layout.view_reference_pager_item, parent, false);
            return new ViewHolder(view);
        }

        @Override
        public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
            ((ViewHolder) holder).bindItem(processLinkTextWithAlphaReferences(references.get(position).getText()),
                    StringUtil.fromHtml(StringUtil.removeCiteMarkup(StringUtil.removeStyleTags(references.get(position).getContent()))));
        }
    }

    @Nullable
    public Callback callback() {
        return FragmentUtil.getCallback(this, Callback.class);
    }
}

File: app/src/main/java/org/wikipedia/page/references/ReferenceDialog.kt
package org.wikipedia.page.references

import android.app.Dialog
import android.graphics.Color
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.view.isVisible
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.tabs.TabLayoutMediator
import org.jsoup.Jsoup
import org.wikipedia.R
import org.wikipedia.WikipediaApp
import org.wikipedia.activity.FragmentUtil.getCallback
import org.wikipedia.bridge.JavaScriptActionHandler
import org.wikipedia.databinding.FragmentReferencesPagerBinding
import org.wikipedia.databinding.ViewReferencePagerItemBinding
import org.wikipedia.dataclient.okhttp.OkHttpWebViewClient
import org.wikipedia.page.ExtendedBottomSheetDialogFragment
import org.wikipedia.page.LinkHandler
import org.wikipedia.page.PageViewModel
import org.wikipedia.util.DimenUtil
import org.wikipedia.util.L10nUtil
import org.wikipedia.util.ResourceUtil
import org.wikipedia.util.StringUtil
import java.util.Locale

class ReferenceDialog : ExtendedBottomSheetDialogFragment() {
    interface Callback {
        val linkHandler: LinkHandler
        val referencesGroup: List<PageReferences.Reference>?
        val selectedReferenceIndex: Int
    }

    private var _binding: FragmentReferencesPagerBinding? = null
    private val binding get() = _binding!!
    private val blankModel = PageViewModel()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentReferencesPagerBinding.inflate(inflater, container, false)
        callback()?.let {
            it.referencesGroup?.run {
                binding.referenceTitleText.text = requireContext().getString(R.string.reference_title, "")
                binding.referencePager.offscreenPageLimit = 2
                binding.referencePager.adapter = ReferencesAdapter(this)
                TabLayoutMediator(binding.pageIndicatorView, binding.referencePager) { _, _ -> }.attach()
                binding.referencePager.setCurrentItem(it.selectedReferenceIndex, true)
                L10nUtil.setConditionalLayoutDirection(binding.root, it.linkHandler.wikiSite.languageCode)
            } ?: return@let null
        } ?: run {
            dismiss()
        }
        return binding.root
    }

    override fun onStart() {
        super.onStart()
        if (callback()?.referencesGroup?.size == 1) {
            binding.pageIndicatorView.visibility = View.GONE
            binding.indicatorDivider.visibility = View.GONE
        }
        BottomSheetBehavior.from(binding.root.parent as View).peekHeight = DimenUtil.displayHeightPx
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    private fun processLinkTextWithAlphaReferences(linkText: String): String {
        var newLinkText = linkText
        val isLowercase = newLinkText.contains("lower")
        if (newLinkText.contains("alpha ")) {
            val strings = newLinkText.split(" ")
            var alphaReference = StringUtil.getBase26String(strings.last().replace("]", "").toInt())
            alphaReference = if (isLowercase) alphaReference.lowercase(Locale.getDefault()) else alphaReference
            newLinkText = alphaReference
        }
        return newLinkText.replace("[\\[\\]]".toRegex(), "") + "."
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        return object : BottomSheetDialog(requireActivity(), theme) {
            override fun onBackPressed() {
                if (binding.referencePager.currentItem > 0) {
                    binding.referencePager.setCurrentItem(binding.referencePager.currentItem - 1, true)
                } else {
                    super.onBackPressed()
                }
            }
        }
    }

    private inner class ViewHolder(val binding: ViewReferencePagerItemBinding) : RecyclerView.ViewHolder(binding.root) {

        fun bindItem(idText: CharSequence?, reference: PageReferences.Reference) {
            binding.referenceId.text = idText
            binding.root.post {
                if (isAdded) {
                    val contents = reference.html
                    if (contents.isEmpty()) {
                        // Inspect html for links without anchor text
                        val links = Jsoup.parse(reference.html).select("a[href]").map { it.attr("href") }
                        var tags = ""
                        for (i in links.indices) {
                            tags = tags.plus("<a href='${links[i]}'>[${i + 1}]</a>")
                        }
                        setContent(tags)
                        binding.referenceExtLink.isVisible = tags.isNotEmpty()
                        return@post
                    }
                    binding.referenceExtLink.isVisible = false
                    setContent(contents)
                }
            }
        }

        private fun setContent(html: String) {
            val wikiSite = callback()?.linkHandler?.wikiSite ?: WikipediaApp.instance.wikiSite
            val colorHex = ResourceUtil.colorToCssString(
                ResourceUtil.getThemedColor(
                    requireContext(),
                    android.R.attr.textColorPrimary
                )
            )
            val dir = if (L10nUtil.isLangRTL(wikiSite.languageCode)) "rtl" else "ltr"
            binding.referenceTextWebView.setBackgroundColor(Color.TRANSPARENT)
            binding.referenceTextWebView.webViewClient = object : OkHttpWebViewClient() {
                override val model get() = blankModel
                override val linkHandler get() = callback()?.linkHandler!!
                override val linkHandlerOverride get() = true
            }
            binding.referenceTextWebView.loadDataWithBaseURL(
                wikiSite.uri.buildUpon().toString(),
                "${JavaScriptActionHandler.getCssStyles(wikiSite)}<div style=\"line-height: 150%; color: #$colorHex\" dir=\"$dir\">$html</div>",
                "text/html",
                "UTF-8",
                null
            )
        }
    }

    private inner class ReferencesAdapter(val references: List<PageReferences.Reference>) : RecyclerView.Adapter<ViewHolder>() {
        override fun getItemCount(): Int {
            return references.size
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            return ViewHolder(ViewReferencePagerItemBinding.inflate(LayoutInflater.from(context), parent, false))
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            holder.bindItem(processLinkTextWithAlphaReferences(references[position].text), references[position])
        }
    }

    private fun callback(): Callback? {
        return getCallback(this, Callback::class.java)
    }
}
